File: peer-fixtures/nocterm/sb4_log_region/bin/sb4_log_region_benchmark.dart
import 'dart:convert';
import 'dart:io';
import 'dart:math' as math;

import 'package:fleury_peer_nocterm_sb4_log_region/log_region_app.dart';
import 'package:nocterm/nocterm.dart';

const _schemaVersion = 1;
const _peerId = 'nocterm';
const _peerName = 'Nocterm';
const _peerVersion = '0.6.0';
const _peerUrl = 'https://pub.dev/packages/nocterm';
const _scenarioId = 'SB.4';
const _defaultWarmups = 2;
const _defaultIterations = 5;
const _defaultRows = 100000;
const _defaultAppend = 1000;
const _defaultSize = Size(120, 32);

Future<void> main(List<String> args) async {
  final options = _Options.parse(args);

  for (var i = 0; i < options.warmupIterations; i += 1) {
    await _runSample(options);
  }

  final samples = <_Sample>[];
  for (var i = 0; i < options.measuredIterations; i += 1) {
    samples.add(await _runSample(options));
  }

  final artifact = _buildArtifact(options, samples);
  final jsonText = const JsonEncoder.withIndent('  ').convert(artifact);
  if (options.outputPath != null) {
    final output = File(options.outputPath!);
    output.parent.createSync(recursive: true);
    output.writeAsStringSync('$jsonText\n');
  }

  if (options.printJson) {
    stdout.writeln(jsonText);
    return;
  }

  final metrics = artifact['metrics']! as Map<String, Object?>;
  stdout.writeln('Nocterm SB.4 log region fixture');
  stdout.writeln('Run: ${artifact['runId']}');
  stdout.writeln('Rows: ${options.rows}');
  stdout.writeln('Append: ${options.appendCount}');
  stdout.writeln('Iterations: ${options.measuredIterations}');
  stdout.writeln(
    'appendBurstUs p95: ${(metrics['appendBurstUs']! as Map)['p95']}',
  );
  stdout.writeln(
    'filterQueryUs p95: ${(metrics['filterQueryUs']! as Map)['p95']}',
  );
  stdout.writeln(
    'unsafeArtifactLeakCount: ${metrics['unsafeArtifactLeakCount']}',
  );
  if (options.outputPath != null) {
    stdout.writeln('Saved ${options.outputPath}');
  }
}

Future<_Sample> _runSample(_Options options) async {
  final rssBefore = ProcessInfo.currentRss;
  final tester = await NoctermTester.create(size: options.size);
  try {
    final mount = Stopwatch()..start();
    await tester.pumpComponent(
      Sb4NoctermLogRegion(
        rowCount: options.rows,
        width: options.size.width.toInt(),
        height: options.size.height.toInt(),
      ),
    );
    mount.stop();
    final state = tester.findState<Sb4NoctermLogRegionState>();

    final firstRender = Stopwatch()..start();
    final firstView = tester.terminalState.getText();
    firstRender.stop();

    final append = Stopwatch()..start();
    state.appendBurst(options.appendCount);
    await tester.pump();
    state.scrollToTail();
    await tester.pump();
    append.stop();
    final appendState = state.snapshot();
    final appendView = tester.terminalState.getText();

    final scrollbackIndex = options.rows ~/ 2;
    final scrollback = Stopwatch()..start();
    state.jumpToScrollback(scrollbackIndex);
    await tester.pump();
    scrollback.stop();
    final scrollbackState = state.snapshot();
    final scrollbackView = tester.terminalState.getText();

    final tail = Stopwatch()..start();
    state.scrollToTail();
    await tester.pump();
    tail.stop();
    final tailState = state.snapshot();
    final tailView = tester.terminalState.getText();

    final copy = Stopwatch()..start();
    state.copySelectedEntry();
    copy.stop();
    final copied = state.lastCopiedText;

    final filter = Stopwatch()..start();
    state.filterQuery(appendFilterQuery());
    await tester.pump();
    state.scrollDisplayedToEnd();
    await tester.pump();
    filter.stop();
    final filterState = state.snapshot();
    final filterView = tester.terminalState.getText();

    final query = Stopwatch()..start();
    final selectedKeyVisible = tester.terminalState.containsText(
      filterState.selectedKey,
    );
    final unsafeLeakCount = filterState.unsafeArtifactLeakCount +
        unsafeVisibleTextCount(firstView) +
        unsafeVisibleTextCount(appendView) +
        unsafeVisibleTextCount(scrollbackView) +
        unsafeVisibleTextCount(tailView) +
        unsafeVisibleTextCount(filterView);
    query.stop();

    final expectedLastIndex = options.rows + options.appendCount - 1;
    final expectedLastKey = logKey(expectedLastIndex);

    return _Sample(
      mountUs: mount.elapsedMicroseconds,
      firstRenderUs: firstRender.elapsedMicroseconds,
      appendBurstUs: append.elapsedMicroseconds,
      scrollbackJumpUs: scrollback.elapsedMicroseconds,
      scrollToTailUs: tail.elapsedMicroseconds,
      copySelectedEntryUs: copy.elapsedMicroseconds,
      filterQueryUs: filter.elapsedMicroseconds,
      semanticOrTestQueryUs: query.elapsedMicroseconds,
      rssDeltaBytes: math.max(0, ProcessInfo.currentRss - rssBefore),
      unsafeArtifactLeakCount: unsafeLeakCount,
      entryCountAfterAppend: appendState.entryCount,
      lineCountAfterFilter: filterState.displayedCount,
      filterMatchCount: filterState.displayedCount,
      selectedKey: filterState.selectedKey,
      scrollY: filterState.scrollY,
      maxScrollY: filterState.maxScrollY,
      visibleWindowRows: filterState.visibleWindowRows,
      tailAnchoringCorrect: appendState.tailAnchored &&
          tailState.tailAnchored &&
          tailState.selectedKey == expectedLastKey,
      copyTextSanitized: copied == expectedCopiedText(expectedLastIndex) &&
          unsafeCopyTextCount(copied) == 0,
      filterResultCorrect: filterState.displayedCount == options.appendCount &&
          filterState.selectedKey == expectedLastKey &&
          selectedKeyVisible,
      scrollbackSelectedCorrect:
          scrollbackState.selectedKey == logKey(scrollbackIndex) &&
              scrollbackView.contains(logKey(scrollbackIndex)),
    );
  } finally {
    tester.dispose();
  }
}

Map<String, Object?> _buildArtifact(_Options options, List<_Sample> samples) {
  final packageRoot = _packageRoot();
  final appLines = _sourceLineCount(
    '${packageRoot.path}/lib/log_region_app.dart',
  );
  final benchmarkLines = _sourceLineCount(
    '${packageRoot.path}/bin/sb4_log_region_benchmark.dart',
  );
  final testLines = _sourceLineCount(
    '${packageRoot.path}/test/log_region_benchmark_test.dart',
  );
  final capturedAt = DateTime.now().toUtc();
  final runId = 'nocterm-sb4-log-region-${_timestampForId(capturedAt)}';
  final unsafeLeakCount =
      samples.map((sample) => sample.unsafeArtifactLeakCount).reduce(math.max);
  final last = samples.last;

  return <String, Object?>{
    'schemaVersion': _schemaVersion,
    'kind': 'fleuryPeerBenchmarkRun',
    'runId': runId,
    'peerId': _peerId,
    'scenarioId': _scenarioId,
    'capturedAt': capturedAt.toIso8601String(),
    'source': <String, Object?>{
      'name': _peerName,
      'version': _peerVersion,
      'url': _peerUrl,
    },
    'environment': <String, Object?>{
      'machine': Platform.localHostname,
      'operatingSystem': Platform.operatingSystem,
      'operatingSystemVersion': Platform.operatingSystemVersion,
      'runtime': Platform.version,
      'terminalMode': 'nocterm-test-harness',
      'terminalSize': <String, Object?>{
        'columns': options.size.width.toInt(),
        'rows': options.size.height.toInt(),
      },
    },
    'fixture': <String, Object?>{
      'workingDirectory': 'peer-fixtures/nocterm/sb4_log_region',
      'command': <String>[
        'dart',
        'run',
        'bin/sb4_log_region_benchmark.dart',
        '--warmup=${options.warmupIterations}',
        '--iterations=${options.measuredIterations}',
        '--rows=${options.rows}',
        '--append=${options.appendCount}',
        '--json',
      ],
      'warmupIterations': options.warmupIterations,
      'measuredIterations': options.measuredIterations,
    },
    'metrics': <String, Object?>{
      'mountUs': _stats(samples.map((sample) => sample.mountUs)),
      'firstRenderUs': _stats(samples.map((sample) => sample.firstRenderUs)),
      'appendBurstUs': _stats(samples.map((sample) => sample.appendBurstUs)),
      'scrollbackJumpUs': _stats(
        samples.map((sample) => sample.scrollbackJumpUs),
      ),
      'scrollToTailUs': _stats(samples.map((sample) => sample.scrollToTailUs)),
      'copySelectedEntryUs': _stats(
        samples.map((sample) => sample.copySelectedEntryUs),
      ),
      'filterQueryUs': _stats(samples.map((sample) => sample.filterQueryUs)),
      'semanticOrTestQueryUs': _stats(
        samples.map((sample) => sample.semanticOrTestQueryUs),
      ),
      'unsafeArtifactLeakCount': unsafeLeakCount,
      'rssDeltaBytes':
          samples.map((sample) => sample.rssDeltaBytes).reduce(math.max),
      'lineOfCodeCount': appLines,
      'benchmarkLineOfCodeCount': benchmarkLines,
      'testLineOfCodeCount': testLines,
      'entryCountAfterAppend': last.entryCountAfterAppend,
      'appendCount': options.appendCount,
      'lineCountAfterFilter': last.lineCountAfterFilter,
      'filterMatchCount': last.filterMatchCount,
      'selectedKey': last.selectedKey,
      'finalScrollY': last.scrollY,
      'finalMaxScrollY': last.maxScrollY,
      'visibleWindowRowEstimate': last.visibleWindowRows,
    },
    'correctness': <Object?>[
      <String, Object?>{
        'gate': 'tail anchoring is correct',
        'pass': samples.every((sample) => sample.tailAnchoringCorrect),
        'evidence':
            'After append and explicit tail scroll, the Nocterm ListView stayed anchored at ${logKey(options.rows + options.appendCount - 1)}.',
      },
      <String, Object?>{
        'gate': 'copy text is sanitized',
        'pass': samples.every((sample) => sample.copyTextSanitized),
        'evidence':
            'Selected-entry copy matched the generated sanitized log line and contained no escape, secret, or newline artifacts.',
      },
      <String, Object?>{
        'gate': 'unsafe output leak count is zero',
        'pass': unsafeLeakCount == 0,
        'evidence':
            'Fixture-owned sanitizer removed ANSI/OSC/control payloads before Nocterm Text/ListView rendering.',
      },
    ],
    'ergonomics': <String, Object?>{
      'lineOfCodeCount': appLines,
      'benchmarkLineOfCodeCount': benchmarkLines,
      'testLineOfCodeCount': testLines,
      'appFile': 'lib/log_region_app.dart',
      'benchmarkFile': 'bin/sb4_log_region_benchmark.dart',
      'testFile': 'test/log_region_benchmark_test.dart',
      'peerOwnedListView': true,
      'peerOwnedScrollController': true,
      'appOwnedSanitization': true,
      'appOwnedFiltering': true,
      'appOwnedSelectedEntryCopy': true,
      'semanticGraphAvailable': false,
      'testQueryViaTerminalStateAndAppState': true,
    },
    'artifacts': <Object?>[
      <String, Object?>{
        'kind': 'source',
        'path': 'peer-fixtures/nocterm/sb4_log_region/lib/log_region_app.dart',
      },
      <String, Object?>{
        'kind': 'benchmark',
        'path':
            'peer-fixtures/nocterm/sb4_log_region/bin/sb4_log_region_benchmark.dart',
      },
      <String, Object?>{
        'kind': 'test',
        'path':
            'peer-fixtures/nocterm/sb4_log_region/test/log_region_benchmark_test.dart',
      },
    ],
    'notes': <String>[
      'This is a Nocterm test-harness peer fixture, not a real-terminal run.',
      'Nocterm 0.6.0 supplies ListView.builder, lazy list rendering, ScrollController, Text rendering, and terminal-state test queries.',
      'Sanitization/redaction, filtering, selected-entry state, and copy/export are app-owned fixture code because Nocterm 0.6.0 does not expose Fleury-equivalent built-in primitives for those SB.4 behaviors.',
      'Nocterm exposes terminal/app state for this fixture, not a Fleury-style semantic app graph.',
      'Timing is local-machine evidence and should not be used as a public superiority claim without repeated peer runs.',
    ],
  };
}

Map<String, Object?> _stats(Iterable<int> values) {
  final sorted = values.toList()..sort();
  if (sorted.isEmpty) {
    throw StateError('Cannot summarize empty metric samples.');
  }
  return <String, Object?>{
    'min': sorted.first,
    'median': _percentile(sorted, 0.50),
    'p95': _percentile(sorted, 0.95),
    'p99': _percentile(sorted, 0.99),
    'max': sorted.last,
    'samples': sorted.length,
  };
}

int _percentile(List<int> sorted, double percentile) {
  final index = ((sorted.length * percentile).ceil() - 1).clamp(
    0,
    sorted.length - 1,
  );
  return sorted[index];
}

Directory _packageRoot() {
  final script = File(Platform.script.toFilePath()).absolute;
  return script.parent.parent;
}

int _sourceLineCount(String path) {
  final lines = File(path).readAsLinesSync();
  var count = 0;
  for (final line in lines) {
    final trimmed = line.trim();
    if (trimmed.isEmpty || trimmed.startsWith('//')) continue;
    count += 1;
  }
  return count;
}

String _timestampForId(DateTime value) {
  return value.toIso8601String().split('.').first.replaceAll(':', '-');
}

final class _Sample {
  const _Sample({
    required this.mountUs,
    required this.firstRenderUs,
    required this.appendBurstUs,
    required this.scrollbackJumpUs,
    required this.scrollToTailUs,
    required this.copySelectedEntryUs,
    required this.filterQueryUs,
    required this.semanticOrTestQueryUs,
    required this.rssDeltaBytes,
    required this.unsafeArtifactLeakCount,
    required this.entryCountAfterAppend,
    required this.lineCountAfterFilter,
    required this.filterMatchCount,
    required this.selectedKey,
    required this.scrollY,
    required this.maxScrollY,
    required this.visibleWindowRows,
    required this.tailAnchoringCorrect,
    required this.copyTextSanitized,
    required this.filterResultCorrect,
    required this.scrollbackSelectedCorrect,
  });

  final int mountUs;
  final int firstRenderUs;
  final int appendBurstUs;
  final int scrollbackJumpUs;
  final int scrollToTailUs;
  final int copySelectedEntryUs;
  final int filterQueryUs;
  final int semanticOrTestQueryUs;
  final int rssDeltaBytes;
  final int unsafeArtifactLeakCount;
  final int entryCountAfterAppend;
  final int lineCountAfterFilter;
  final int filterMatchCount;
  final String selectedKey;
  final int scrollY;
  final int maxScrollY;
  final int visibleWindowRows;
  final bool tailAnchoringCorrect;
  final bool copyTextSanitized;
  final bool filterResultCorrect;
  final bool scrollbackSelectedCorrect;
}

final class _Options {
  const _Options({
    required this.warmupIterations,
    required this.measuredIterations,
    required this.rows,
    required this.appendCount,
    required this.size,
    required this.printJson,
    required this.outputPath,
  });

  final int warmupIterations;
  final int measuredIterations;
  final int rows;
  final int appendCount;
  final Size size;
  final bool printJson;
  final String? outputPath;

  static _Options parse(List<String> args) {
    var warmupIterations = _defaultWarmups;
    var measuredIterations = _defaultIterations;
    var rows = _defaultRows;
    var appendCount = _defaultAppend;
    var size = _defaultSize;
    var printJson = false;
    String? outputPath;

    for (final arg in args) {
      if (arg == '--json') {
        printJson = true;
      } else if (arg.startsWith('--warmup=')) {
        warmupIterations = _parseNonNegativeInt(arg, '--warmup=');
      } else if (arg.startsWith('--iterations=')) {
        measuredIterations = _parsePositiveInt(arg, '--iterations=');
      } else if (arg.startsWith('--rows=')) {
        rows = _parsePositiveInt(arg, '--rows=');
      } else if (arg.startsWith('--append=')) {
        appendCount = _parsePositiveInt(arg, '--append=');
      } else if (arg.startsWith('--size=')) {
        size = _parseSize(arg.substring('--size='.length));
      } else if (arg.startsWith('--output=')) {
        outputPath = arg.substring('--output='.length);
      } else {
        throw ArgumentError('Unknown argument: $arg');
      }
    }

    return _Options(
      warmupIterations: warmupIterations,
      measuredIterations: measuredIterations,
      rows: rows,
      appendCount: appendCount,
      size: size,
      printJson: printJson,
      outputPath: outputPath,
    );
  }

  static int _parsePositiveInt(String arg, String prefix) {
    final value = int.tryParse(arg.substring(prefix.length));
    if (value == null || value <= 0) {
      throw ArgumentError('$prefix requires a positive integer.');
    }
    return value;
  }

  static int _parseNonNegativeInt(String arg, String prefix) {
    final value = int.tryParse(arg.substring(prefix.length));
    if (value == null || value < 0) {
      throw ArgumentError('$prefix requires a non-negative integer.');
    }
    return value;
  }

  static Size _parseSize(String value) {
    final parts = value.split('x');
    if (parts.length != 2) {
      throw ArgumentError('--size requires COLUMNSxROWS.');
    }
    final columns = int.tryParse(parts[0]);
    final rows = int.tryParse(parts[1]);
    if (columns == null || rows == null || columns <= 0 || rows <= 0) {
      throw ArgumentError('--size requires positive integer dimensions.');
    }
    return Size(columns.toDouble(), rows.toDouble());
  }
}

File: peer-fixtures/opentui/sb4_log_region/bin/sb4_log_region_benchmark.js
#!/usr/bin/env bun

import { execFileSync } from 'node:child_process'
import { readFileSync, mkdirSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

import {
  Sb4OpenTuiLogHarness,
  appendFilterQuery,
  expectedCopiedText,
  logKey,
  unsafeCopyTextCount,
  unsafeVisibleTextCount,
} from '../lib/log_app.js'

const SCHEMA_VERSION = 1
const PEER_ID = 'opentui'
const PEER_NAME = 'OpenTUI'
const PEER_VERSION = '0.3.1'
const PEER_URL = 'https://www.npmjs.com/package/@opentui/core'
const SCENARIO_ID = 'SB.4'
const DEFAULT_WARMUPS = 2
const DEFAULT_ITERATIONS = 5
const DEFAULT_ROWS = 100_000
const DEFAULT_APPEND = 1_000
const DEFAULT_COLUMNS = 120
const DEFAULT_TERMINAL_ROWS = 32

const options = parseArgs(process.argv.slice(2))

for (let index = 0; index < options.warmupIterations; index += 1) {
  await runSample(options)
}

const rssBefore = process.memoryUsage().rss
const samples = []
for (let index = 0; index < options.measuredIterations; index += 1) {
  samples.push(await runSample(options))
}
const runRssDeltaBytes = Math.max(0, process.memoryUsage().rss - rssBefore)
const artifact = buildArtifact(options, samples, runRssDeltaBytes)
const jsonText = `${JSON.stringify(artifact, null, 2)}\n`

if (options.outputPath) {
  mkdirSync(dirname(options.outputPath), { recursive: true })
  writeFileSync(options.outputPath, jsonText)
}

if (options.printJson) {
  process.stdout.write(jsonText)
} else {
  console.log('OpenTUI SB.4 LogRegion fixture')
  console.log(`Rows: ${options.rows}`)
  console.log(`Append: ${options.appendCount}`)
  console.log(`Iterations: ${options.measuredIterations}`)
  console.log(`appendBurstUs p95: ${artifact.metrics.appendBurstUs.p95}`)
  console.log(`filterQueryUs p95: ${artifact.metrics.filterQueryUs.p95}`)
  console.log(`unsafeArtifactLeakCount: ${artifact.metrics.unsafeArtifactLeakCount}`)
  if (options.outputPath) console.log(`Saved ${options.outputPath}`)
}

async function runSample(options) {
  let app
  const rssBefore = process.memoryUsage().rss
  const mountStart = process.hrtime.bigint()
  app = await Sb4OpenTuiLogHarness.create({
    rowCount: options.rows,
    terminalColumns: options.terminalColumns,
    terminalRows: options.terminalRows,
  })
  const mountUs = elapsedUs(mountStart)
  const rssAfterMount = process.memoryUsage().rss

  try {
    const firstRenderStart = process.hrtime.bigint()
    const firstFrame = await app.render()
    const firstRenderUs = elapsedUs(firstRenderStart)

    const appendStart = process.hrtime.bigint()
    app.appendBurst(options.appendCount)
    const appendFrame = await app.render()
    const appendBurstUs = elapsedUs(appendStart)
    const appendState = app.snapshot()

    const scrollbackIndex = Math.floor(options.rows / 2)
    const scrollbackStart = process.hrtime.bigint()
    app.jumpToScrollback(scrollbackIndex)
    const scrollbackFrame = await app.render()
    const scrollbackJumpUs = elapsedUs(scrollbackStart)
    const scrollbackState = app.snapshot()

    const tailStart = process.hrtime.bigint()
    app.scrollToTail()
    const tailFrame = await app.render()
    const scrollToTailUs = elapsedUs(tailStart)
    const tailState = app.snapshot()

    const copyStart = process.hrtime.bigint()
    const copied = app.copySelectedEntry()
    const copySelectedEntryUs = elapsedUs(copyStart)

    const filterStart = process.hrtime.bigint()
    app.filterQuery(appendFilterQuery())
    const filterFrame = await app.render()
    const filterQueryUs = elapsedUs(filterStart)
    const filterState = app.snapshot()

    const queryStart = process.hrtime.bigint()
    const selectedKeyVisible = filterState.frameContainsSelected
    const unsafeLeakCount =
      filterState.lastFrameUnsafeCount +
      unsafeCopyTextCount(copied) +
      unsafeVisibleTextCount(firstFrame) +
      unsafeVisibleTextCount(appendFrame) +
      unsafeVisibleTextCount(scrollbackFrame) +
      unsafeVisibleTextCount(tailFrame) +
      unsafeVisibleTextCount(filterFrame)
    const semanticOrTestQueryUs = elapsedUs(queryStart)

    const expectedLastIndex = options.rows + options.appendCount - 1
    const expectedLastKey = logKey(expectedLastIndex)

    return {
      mountUs,
      firstRenderUs,
      appendBurstUs,
      scrollbackJumpUs,
      scrollToTailUs,
      copySelectedEntryUs,
      filterQueryUs,
      semanticOrTestQueryUs,
      rssDeltaBytes: Math.max(0, rssAfterMount - rssBefore),
      unsafeArtifactLeakCount: unsafeLeakCount,
      entryCountAfterAppend: appendState.entryCount,
      lineCountAfterFilter: filterState.displayedCount,
      filterMatchCount: filterState.displayedCount,
      selectedKey: filterState.selectedKey,
      visibleStart: filterState.visibleStart,
      visibleEnd: filterState.visibleEnd,
      visibleWindowRows: filterState.visibleWindowRows,
      tailAnchoringCorrect:
        appendState.tailAnchored &&
        tailState.tailAnchored &&
        tailState.selectedKey === expectedLastKey &&
        tailState.entryCount === options.rows + options.appendCount,
      copyTextSanitized:
        copied === expectedCopiedText(expectedLastIndex, 'append') &&
        unsafeCopyTextCount(copied) === 0,
      filterResultCorrect:
        filterState.displayedCount === options.appendCount &&
        filterState.selectedKey === expectedLastKey &&
        selectedKeyVisible,
      scrollbackSelectedCorrect:
        scrollbackState.selectedKey === logKey(scrollbackIndex) &&
        scrollbackFrame.includes(logKey(scrollbackIndex)),
    }
  } finally {
    app?.destroy()
  }
}

function buildArtifact(options, samples, runRssDeltaBytes) {
  const capturedAt = new Date()
  const runId = `opentui-sb4-log-region-${capturedAt
    .toISOString()
    .replace(/\.\d{3}Z$/, 'Z')
    .replaceAll(':', '-')}`
  const last = samples[samples.length - 1]
  const unsafeLeakCount = Math.max(
    ...samples.map((sample) => sample.unsafeArtifactLeakCount),
  )
  const appLines = sourceLineCount(new URL('../lib/log_app.js', import.meta.url))
  const benchmarkLines = sourceLineCount(import.meta.url)
  const testLines = sourceLineCount(new URL('../test/log_app.test.js', import.meta.url))

  return {
    schemaVersion: SCHEMA_VERSION,
    kind: 'fleuryPeerBenchmarkRun',
    runId,
    peerId: PEER_ID,
    scenarioId: SCENARIO_ID,
    capturedAt: capturedAt.toISOString(),
    source: {
      name: PEER_NAME,
      version: PEER_VERSION,
      url: PEER_URL,
    },
    environment: {
      machine: commandOutput('hostname') ?? 'unknown',
      operatingSystem: process.platform,
      operatingSystemVersion:
        commandOutput('uname', ['-a']) ?? process.platform,
      runtime: `${runtimeVersion()} / OpenTUI ${PEER_VERSION}`,
      terminalMode: 'opentui-test-renderer-memory',
      terminalSize: {
        columns: options.terminalColumns,
        rows: options.terminalRows,
      },
    },
    fixture: {
      workingDirectory: 'peer-fixtures/opentui/sb4_log_region',
      command: [
        'npm',
        'run',
        'benchmark',
        '--',
        `--warmup=${options.warmupIterations}`,
        `--iterations=${options.measuredIterations}`,
        `--rows=${options.rows}`,
        `--append=${options.appendCount}`,
        '--json',
      ],
      warmupIterations: options.warmupIterations,
      measuredIterations: options.measuredIterations,
    },
    metrics: {
      mountUs: stats(samples.map((sample) => sample.mountUs)),
      firstRenderUs: stats(samples.map((sample) => sample.firstRenderUs)),
      appendBurstUs: stats(samples.map((sample) => sample.appendBurstUs)),
      scrollbackJumpUs: stats(samples.map((sample) => sample.scrollbackJumpUs)),
      scrollToTailUs: stats(samples.map((sample) => sample.scrollToTailUs)),
      copySelectedEntryUs: stats(
        samples.map((sample) => sample.copySelectedEntryUs),
      ),
      filterQueryUs: stats(samples.map((sample) => sample.filterQueryUs)),
      semanticOrTestQueryUs: stats(
        samples.map((sample) => sample.semanticOrTestQueryUs),
      ),
      unsafeArtifactLeakCount: unsafeLeakCount,
      rssDeltaBytes: Math.max(
        runRssDeltaBytes,
        ...samples.map((sample) => sample.rssDeltaBytes),
      ),
      lineOfCodeCount: appLines,
      benchmarkLineOfCodeCount: benchmarkLines,
      testLineOfCodeCount: testLines,
      entryCountAfterAppend: last.entryCountAfterAppend,
      appendCount: options.appendCount,
      lineCountAfterFilter: last.lineCountAfterFilter,
      filterMatchCount: last.filterMatchCount,
      selectedKey: last.selectedKey,
      finalVisibleStart: last.visibleStart,
      finalVisibleEnd: last.visibleEnd,
      visibleWindowRowEstimate: last.visibleWindowRows,
    },
    correctness: [
      {
        gate: 'tail anchoring is correct',
        pass: samples.every((sample) => sample.tailAnchoringCorrect),
        evidence: `After append and explicit tail scroll, the OpenTUI text frame stayed anchored at ${logKey(
          options.rows + options.appendCount - 1,
        )}.`,
      },
      {
        gate: 'copy text is sanitized',
        pass: samples.every((sample) => sample.copyTextSanitized),
        evidence:
          'Selected-entry copy matched the generated sanitized log line and contained no escape, secret, or newline artifacts.',
      },
      {
        gate: 'unsafe output leak count is zero',
        pass: unsafeLeakCount === 0,
        evidence:
          'Fixture-owned sanitizer removed ANSI/OSC/control payloads before OpenTUI TextRenderable rendering.',
      },
    ],
    ergonomics: {
      lineOfCodeCount: appLines,
      benchmarkLineOfCodeCount: benchmarkLines,
      testLineOfCodeCount: testLines,
      appFile: 'lib/log_app.js',
      benchmarkFile: 'bin/sb4_log_region_benchmark.js',
      testFile: 'test/log_app.test.js',
      peerOwnedTextRenderable: true,
      peerOwnedTestRenderer: true,
      appOwnedRetainedLogs: true,
      appOwnedTailPolicy: true,
      appOwnedSanitization: true,
      appOwnedFiltering: true,
      appOwnedSelectedEntryCopy: true,
      semanticGraphAvailable: false,
      testQueryViaFrameAndAppState: true,
    },
    artifacts: [
      {
        kind: 'source',
        path: 'peer-fixtures/opentui/sb4_log_region/lib/log_app.js',
      },
      {
        kind: 'benchmark',
        path:
          'peer-fixtures/opentui/sb4_log_region/bin/sb4_log_region_benchmark.js',
      },
      {
        kind: 'test',
        path: 'peer-fixtures/opentui/sb4_log_region/test/log_app.test.js',
      },
    ],
    notes: [
      'This is an OpenTUI test-renderer peer fixture, not a real-terminal run.',
      'OpenTUI 0.3.1 supplies TextRenderable and the native-backed test renderer.',
      'Retained logs, tail policy, scrollback selection, sanitization/redaction, filtering, selected-entry state, and copy/export are app-owned fixture code because OpenTUI 0.3.1 does not expose Fleury-equivalent built-in LogRegion semantics.',
      'OpenTUI exposes frame/app state for this fixture, not a Fleury-style semantic app graph.',
      'Timing is local-machine evidence and should not be used as a public superiority claim without repeated peer runs.',
    ],
  }
}

function stats(values) {
  const sorted = [...values].sort((a, b) => a - b)
  if (sorted.length === 0) throw new Error('Cannot summarize empty metric samples.')
  return {
    min: sorted[0],
    median: percentile(sorted, 0.5),
    p95: percentile(sorted, 0.95),
    p99: percentile(sorted, 0.99),
    max: sorted[sorted.length - 1],
    samples: sorted.length,
  }
}

function percentile(sorted, percent) {
  const index = Math.min(
    sorted.length - 1,
    Math.max(0, Math.ceil(sorted.length * percent) - 1),
  )
  return sorted[index]
}

function elapsedUs(start) {
  return Number((process.hrtime.bigint() - start) / 1000n)
}

function sourceLineCount(urlOrPath) {
  const path =
    urlOrPath instanceof URL || String(urlOrPath).startsWith('file:')
      ? fileURLToPath(urlOrPath)
      : urlOrPath
  return readFileSync(path, 'utf8')
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0 && !line.startsWith('//')).length
}

function commandOutput(command, args = []) {
  try {
    return execFileSync(command, args, { encoding: 'utf8' }).trim()
  } catch (_) {
    return null
  }
}

function runtimeVersion() {
  return commandOutput('bun', ['--version']) ?? process.version
}

function parseArgs(args) {
  const options = {
    warmupIterations: DEFAULT_WARMUPS,
    measuredIterations: DEFAULT_ITERATIONS,
    rows: DEFAULT_ROWS,
    appendCount: DEFAULT_APPEND,
    terminalColumns: DEFAULT_COLUMNS,
    terminalRows: DEFAULT_TERMINAL_ROWS,
    printJson: false,
    outputPath: '',
  }

  for (const arg of args) {
    if (arg === '--json') {
      options.printJson = true
      continue
    }
    const [name, value] = arg.split('=', 2)
    switch (name) {
      case '--warmup':
        options.warmupIterations = parseNonNegativeInt(name, value)
        break
      case '--iterations':
        options.measuredIterations = parsePositiveInt(name, value)
        break
      case '--rows':
        options.rows = parsePositiveInt(name, value)
        break
      case '--append':
        options.appendCount = parsePositiveInt(name, value)
        break
      case '--columns':
        options.terminalColumns = parsePositiveInt(name, value)
        break
      case '--terminal-rows':
        options.terminalRows = parsePositiveInt(name, value)
        break
      case '--output':
        if (!value) throw new Error('--output requires a path')
        options.outputPath = value
        break
      default:
        throw new Error(`Unknown argument: ${arg}`)
    }
  }

  return options
}

function parsePositiveInt(name, value) {
  const parsed = Number.parseInt(value, 10)
  if (!Number.isFinite(parsed) || parsed <= 0) {
    throw new Error(`${name} must be a positive integer.`)
  }
  return parsed
}

function parseNonNegativeInt(name, value) {
  const parsed = Number.parseInt(value, 10)
  if (!Number.isFinite(parsed) || parsed < 0) {
    throw new Error(`${name} must be a non-negative integer.`)
  }
  return parsed
}
